SATSubsumption/subsat/SubstitutionTheory.hpp

```diff
@@ -0,0 +1,399 @@
+        : m_bindings_storage{bindings_storage}
+        , m_bindings_start{bindings_storage.size()}
+    { }
+
+    Binder(Binder&) = delete;
+    Binder& operator=(Binder&) = delete;
+    Binder(Binder&&) = default;
+    Binder& operator=(Binder&&) = delete;
+
+    ~Binder() noexcept
+    {
+      if (m_state == State::Active) {
+        reset();
+      }
+    }
+
+    bool bind(VampireVar var, VampireTerm term)
+    {
+      ASS(m_state == State::Active);
+      for (auto it = m_bindings_storage.cbegin() + m_bindings_start; it != m_bindings_storage.cend(); ++it) {
+        if (it->first == var) {
+          // 'var' is already bound => binding succeeds iff we bind to the same term again
+          return (it->second == term);
+        }
+      }
+      // 'var' is not bound yet => store new binding
+      m_bindings_storage.emplace_back(var, term);
+      return true;
+    }
+
+    void specVar(VampireVar /* var */, VampireTerm /* term */)
+    {
+      ASSERTION_VIOLATION;
+    }
+
+    void reset() noexcept
+    {
+      ASS(m_state == State::Active);
+      ASS_GE(m_bindings_storage.size(), m_bindings_start);
+      m_bindings_storage.resize(m_bindings_start);  // will not throw because we shrink the size
+      m_state = State::Reset;
+    }
+
+    size_t size() const noexcept
+    {
+      ASS_GE(m_bindings_storage.size(), m_bindings_start);
+      return m_bindings_storage.size() - m_bindings_start;
+    }
+
+  private:
+    size_t index() const noexcept
+    {
+      return m_bindings_start;
+    }
+
+    /// Users should call BindingsManager::commit_bindings() instead.
+    void commit() noexcept
+    {
+      ASS(m_state == State::Active);
+      m_state = State::Committed;
+    }
+
+    friend class BindingsManager;
+
+  private:
+    BindingsStorage& m_bindings_storage;
+    /// First index of the current bindings in m_bindings_storage
+    size_t m_bindings_start;
+    /// Current state of the binder
+    State m_state = State::Active;
+  };  // class Binder
+
+  /// Only one binder may be active at any time.
+  Binder start_binder() noexcept
+  {
+    return {m_bindings_storage};
+  }
+
+  /// Commit bindings to storage.
+  void commit_bindings(Binder& binder, subsat::Var b, uint32_t extra_i = 0, uint32_t extra_j = 0)
+  {
+    binder.commit();
+    while (b.index() >= m_bindings.size()) {
+      m_bindings.emplace_back();
+    }
+    BindingsRef& bindings = m_bindings[b];
+    ASS_REP(!bindings.is_valid(), "Bindings for b are already set");
+    bindings.index = binder.index();
+    bindings.size = binder.size();
+    bindings.extra_i = extra_i;
+    bindings.extra_j = extra_j;
+    ASS(bindings.is_valid());
+  }
+
+  struct BindingsRef final
+  {
+    uint32_t index = std::numeric_limits<uint32_t>::max();
+    uint32_t size;
+    uint32_t extra_i;  // NOTE: we don't actually use the extra_i field. But including it seems to increase performance, at least on ARM64 (Apple M1). Because of alignment 16 vs 12?
```

Comment:
Actually this comment was very outdated... we use neither `extra_i` nor `extra_j` anymore, so I removed both. I think this is because we now have the `MatchSet` taking care of this information.